tianshou/utils/space_info.py

```diff
@@ -0,0 +1,111 @@
+from collections.abc import Sequence
+from dataclasses import dataclass
+from typing import Any, Self
+
+import numpy as np
+from gymnasium import spaces
+
+from tianshou.utils.string import ToStringMixin
+
+
+@dataclass(kw_only=True)
+class ActionSpaceInfo(ToStringMixin):
+    """A data structure for storing the different attributes of the action space."""
+
+    action_shape: int | Sequence[int]
+    """The shape of the action space."""
+    min_action: float
+    """The smallest allowable action or in the continuous case the lower bound for allowable action value."""
+    max_action: float
+    """The largest allowable action or in the continuous case the upper bound for allowable action value."""
+
+    @property
+    def action_dim(self) -> int:
+        """Return the number of distinct actions an agent can take it its action space."""
+        if isinstance(self.action_shape, int):
+            return self.action_shape
+        elif isinstance(self.action_shape, Sequence) and self.action_shape:
+            return int(np.prod(self.action_shape))
+        else:
```

Comment:
We're almost there. Accessing a property should never raise an Exception, it's highly unexpected from a user perspective.

In fact, checking for None is not necessary - the type annotation only allows for int and Sequence[int], so nothing can go wrong. Pls remove the checks and the error. Note that we have to trust people passing the annotated types, doesn't make sense to check all of them at runtime. Here an proper IDE or mypy would tell the user if they passed an wrong type for action_shape

NB: if you wanted to raise an error here, it would be a RuntimeError, since it's not caused by user-provided values to the call.